crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -0,0 +1,441 @@
+    /// let file_a = ttd.file("file_a");
+    /// let file_b = ttd.file("file_b");
+    ///
+    /// let mut fs_notification_stream = NotifyStream::try_default().unwrap();
+    /// fs_notification_stream.add_watcher(ttd.path(), String::from("file_a"), &[Masks::Modified]).unwrap();
+    /// fs_notification_stream.add_watcher(ttd.path(), String::from("file_b"), &[Masks::Created, Masks::Deleted]).unwrap();
+    /// ```
+    /// NOTE:
+    /// in this last example, the root directory is the same: `ttd.path()`
+    /// but the files watched and masks are different. In the background,
+    /// the `add_watcher` fn will add a watch on `ttd.path()` with masks:
+    /// Created, Modified and Deleted. and will update `self.watchers`
+    /// with two entries, one for file_a and one for file_b.
+    ///
+    /// The `stream` method will check that events coming from
+    /// `ttd.path()` match `self.watchers`
+    pub fn add_watcher(
+        &mut self,
+        dir_path: &Path,
+        file: String,
+        masks: &[Masks],
+    ) -> Result<(), NotifyStreamError> {
+        // the first step if to normalise `dir_path` and `file`
+        // this step is done just in case `file` is not just the name
+        // but is a partial path to the file. For more info see the
+        // function's docstring.
+        let (dir_path, file) = normalising_watch_dir_and_file(dir_path, &file)?;
+        let dir_path = dir_path.as_path();
+
+        if let Some(mut watcher) = self.watchers.clone() {
+            let mut all_flags: HashSet<_> = watcher
+                .iter()
+                .flat_map(|set| set.key.values())
+                .flatten()
+                .collect();
+            for mask in masks {
+                all_flags.insert(mask);
+            }
+
+            let watch_mask = pipe_masks_into_watch_mask(masks);
+            let _ = self.inotify.add_watch(dir_path, watch_mask);
+            let wd = WatchDescriptor::new(dir_path.to_path_buf(), file, masks.to_vec());
+            watcher.insert(wd);
+            self.watchers = Some(watcher);
+        } else {
+            let watch_mask = pipe_masks_into_watch_mask(masks);
+            let _ = self.inotify.add_watch(dir_path, watch_mask);
+            let wd = WatchDescriptor::new(dir_path.to_path_buf(), file, masks.to_vec());
+            let mut hs = HashSet::new();
+            hs.insert(wd);
+            self.watchers = Some(hs);
+        }
+        Ok(())
+    }
+
+    /// create an fs notification event stream
+    pub fn stream(mut self) -> impl Stream<Item = Result<(PathBuf, Masks), NotifyStreamError>> {
+        try_stream! {
+            let mut notify_service = self.inotify.event_stream(self.buffer)?;
+            while let Some(event_or_error) = notify_service.next().await {
+                match event_or_error {
+                    Ok(event) => {
+                        let event_mask: Masks = event.mask.into();
```

Comment:
This is where I would filter out the unexpected events - making useless the `Masks::Undefined` case.